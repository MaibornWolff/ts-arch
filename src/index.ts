export * from "./core/Project"
export * from "./core/Result"
export * from "./core/Rule"
export * from "./core/TSArch"
export * from "./jest/ArchMatchers"
