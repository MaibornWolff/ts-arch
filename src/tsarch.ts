import { EntryPipe } from './api/core/pipe/EntryPipe'
import { TypescriptProjectParser } from './parser/TypescriptProjectParser'
import { FileSubject } from './api/core/subject/FileSubject'
import { ArchProject } from './api/core/ArchProject'

export class TSArch {
  static defineThat(): EntryPipe {
    return new EntryPipe()
  }

  static async parseTypescriptProject(path: string): Promise<ArchProject> {
    const parser = new TypescriptProjectParser()
    const files = await parser.parse(path)
    const project = new ArchProject()
    files.forEach(file => {
      const fileSubject = new FileSubject(file.parsedPath.base, file.parsedPath.dir)
      project.addSubject(fileSubject)
    })
    return project
  }
}
