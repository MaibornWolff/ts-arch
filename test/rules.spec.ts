import { TSArch } from "../src/tsarch"
import { generateAnimalSubjectsMock } from "./generators"

describe("Basic Rules", () => {
	let subjects

	beforeEach(() => {
		subjects = generateAnimalSubjectsMock()
	})

	it("all files should have no subjects is false", () => {
		const rule = TSArch.defineThat()
			.files()
			.should()
			.not()
			.haveSubjects()
		expect(rule.check(subjects)).toBeFalsy()
	})

	it("all subjects should have no subjects is false", () => {
		const rule = TSArch.defineThat()
			.all()
			.should()
			.not()
			.haveSubjects()
		expect(rule.check(subjects)).toBeFalsy()
	})

	it("all subjects should have subjects is true", () => {
		const rule = TSArch.defineThat()
			.all()
			.should()
			.haveSubjects()
		expect(rule.check(subjects)).toBeTruthy()
	})

	it("all subjects with name matching x should have subjects is false", () => {
		const rule = TSArch.defineThat()
			.all()
			.withNameMatching(/x/)
			.should()
			.haveSubjects()
		expect(rule.check(subjects)).toBeFalsy()
	})

	it("all subjects with name matching .* should have subjects is true", () => {
		const rule = TSArch.defineThat()
			.all()
			.withNameMatching(/.*/)
			.should()
			.haveSubjects()
		expect(rule.check(subjects)).toBeTruthy()
	})

	it("should find all controllers", () => {
		const pipe = TSArch.defineThat()
			.all()
			.withNameMatching(/Controller/)
		expect(pipe.filterSubjects(subjects)).toMatchSnapshot()
	})

	it("should find all Dog stuff", () => {
		const pipe = TSArch.defineThat()
			.all()
			.withNameMatching(/Dog/)
		expect(pipe.filterSubjects(subjects)).toMatchSnapshot()
	})

	it("should find all Services by suffix", () => {
		const pipe = TSArch.defineThat()
			.all()
			.withNameSuffix("Service")
		expect(pipe.filterSubjects(subjects)).toMatchSnapshot()
	})

	it("should find all Dog things by prefix", () => {
		const pipe = TSArch.defineThat()
			.all()
			.withNamePrefix("Dog")
		expect(pipe.filterSubjects(subjects)).toMatchSnapshot()
	})

	it("should not find all Dog things by suffix", () => {
		const pipe = TSArch.defineThat()
			.all()
			.withNameSuffix("Dog")
		expect(pipe.filterSubjects(subjects)).toMatchSnapshot()
	})

	it("should not find all Services by prefix", () => {
		const pipe = TSArch.defineThat()
			.all()
			.withNamePrefix("Service")
		expect(pipe.filterSubjects(subjects)).toMatchSnapshot()
	})

	it("should find all Services by file suffix", () => {
		const pipe = TSArch.defineThat()
			.files()
			.withNameSuffix("Service")
		expect(pipe.filterSubjects(subjects)).toMatchSnapshot()
	})

	it("should find all Services by matching path", () => {
		const pipe = TSArch.defineThat()
			.files()
			.withPathMatching(/.*service.*/)
		expect(pipe.filterSubjects(subjects)).toMatchSnapshot()
	})
})
