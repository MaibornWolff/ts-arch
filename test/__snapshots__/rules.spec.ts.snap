// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic Rules should find all Dog stuff 1`] = `
Array [
  FileSubject {
    "dir": "C://project/src/service",
    "name": "DogService",
  },
  FileSubject {
    "dir": "C://project/src/controller",
    "name": "DogController",
  },
  FileSubject {
    "dir": "C://project/src/model",
    "name": "DogModel",
  },
]
`;

exports[`Basic Rules should find all Dog things by prefix 1`] = `
Array [
  FileSubject {
    "dir": "C://project/src/service",
    "name": "DogService",
  },
  FileSubject {
    "dir": "C://project/src/controller",
    "name": "DogController",
  },
  FileSubject {
    "dir": "C://project/src/model",
    "name": "DogModel",
  },
]
`;

exports[`Basic Rules should find all Services by file suffix 1`] = `
Array [
  FileSubject {
    "dir": "C://project/src/service",
    "name": "DogService",
  },
]
`;

exports[`Basic Rules should find all Services by matching path 1`] = `
Array [
  FileSubject {
    "dir": "C://project/src/service",
    "name": "DogService",
  },
]
`;

exports[`Basic Rules should find all Services by suffix 1`] = `
Array [
  FileSubject {
    "dir": "C://project/src/service",
    "name": "DogService",
  },
]
`;

exports[`Basic Rules should find all controllers 1`] = `
Array [
  FileSubject {
    "dir": "C://project/src/controller",
    "name": "DogController",
  },
  FileSubject {
    "dir": "C://project/src/controller",
    "name": "CatController",
  },
]
`;

exports[`Basic Rules should not find all Dog things by suffix 1`] = `Array []`;

exports[`Basic Rules should not find all Services by prefix 1`] = `Array []`;
